dockstore-webservice/src/main/java/io/dockstore/webservice/core/ToolTester/ToolTesterS3Client.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    /**
+     * Converts the toolId into a key for s3 storage.  Used by both webservice and tooltester
+     * Workflows will be in a "workflow" directory whereas tools will be in a "tool" directory
+     * repository and optional toolname or workflowname must be encoded or else looking for logs of a specific tool without toolname (quay.io/dockstore/hello_world)
+     * will return logs for the other ones with toolnames (quay.io/dockstore/hello_world/thing)
+     * TODO: Somehow reuse this between repos
+     *
+     * @param toolId TRS tool ID
+     * @return The key for s3
+     */
+    static String convertToolIdToPartialKey(String toolId) throws UnsupportedEncodingException {
+        if (toolId.startsWith("#workflow")) {
+            toolId = toolId.replaceFirst("#workflow", "workflow");
+        } else {
+            toolId = "tool/" + toolId;
+        }
+        String[] split = toolId.split("/");
+        if (split.length == MAX_TOOL_ID_STRING_SEGMENTS) {
+            split[TOOL_ID_REPOSITORY_INDEX] = URLEncoder
+                    .encode(split[TOOL_ID_REPOSITORY_INDEX] + "/" + split[TOOL_ID_TOOLNAME_INDEX], StandardCharsets.UTF_8.name());
+            String[] encodedToolIdArray = Arrays.copyOf(split, split.length - 1);
+            return String.join("/", encodedToolIdArray);
+        } else {
+            return toolId;
+        }
+    }
+
+    public String getToolTesterLog(String toolId, String versionName, String testFilePath, String runner, String filename)
+            throws IOException {
+        String key = generateKey(toolId, versionName, testFilePath, runner, filename);
+        S3Object s3Object = s3.getObject(bucketName, key);
+        InputStream objectContent = s3Object.getObjectContent();
+        return IOUtils.toString(objectContent, StandardCharsets.UTF_8);
+
+    }
+
+    public List<ToolTesterLog> getToolTesterLogs(String toolId, String toolVersionName) throws UnsupportedEncodingException {
+        String key = convertToolIdToPartialKey(toolId) + "/" + URLEncoder.encode(toolVersionName, StandardCharsets.UTF_8.name());
+        ObjectListing listing = s3.listObjects(bucketName, key);
+        List<S3ObjectSummary> summaries = listing.getObjectSummaries();
+        while (listing.isTruncated()) {
+            listing = s3.listNextBatchOfObjects(listing);
+            summaries.addAll(listing.getObjectSummaries());
+        }
+        return convertObjectListingToTooltesterLogs(listing);
+    }
+
+    private List<ToolTesterLog> convertObjectListingToTooltesterLogs(ObjectListing listing) {
+        List<ToolTesterLog> toolTesterLogs = new ArrayList<>();
+        List<S3ObjectSummary> summaries = listing.getObjectSummaries();
+        while (listing.isTruncated()) {
+            listing = s3.listNextBatchOfObjects(listing);
+            summaries.addAll(listing.getObjectSummaries());
+        }
+
+        summaries.forEach(summary -> {
```

Comment:
The more functional way would be to `return summaries.map(...`. Just FYI -- I'm still learning it and not a purist, and it's not part of our coding standard.